### PR TITLE
staging-next 2023-12-20

### PR DESCRIPTION
**Pull request by vcunat**
##### Helpful links

- workflow docs: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#staging
- important jobs:
  https://hydra.nixos.org/job/nixpkgs/staging-next/unstable#tabs-constituents
  https://hydra.nixos.org/job/nixos/staging-next-small/tested
- jobsets:
  https://hydra.nixos.org/jobset/nixpkgs/staging-next
  https://hydra.nixos.org/jobset/nixos/staging-next-small
- nix-review reports: https://malob.github.io/nix-review-tools-reports/
- previous staging-next: https://github.com/NixOS/nixpkgs/pull/271377
- - - 
##### Mass breakages
(will be edited based on progress)

- libxml2 reorganized headers. The change was mostly reverted but some patches might be still needed to be cherry-picked from https://github.com/NixOS/nixpkgs/pull/269060

### CODE REVIEW

**Comment by fabianhjr** [comment]
As a heads up there is a default gcc upgrade in this cycle: https://github.com/NixOS/nixpkgs/commit/59048500b27505ef81b13bc0716a2822be616b4e

**Comment by mweinelt** [comment]
And I also reported a number of regressions a week ago in #268097

**Comment by jtojnar** [comment]
@happysalada @natsukium looks like `python311Packages.openllm-core` is broken here (even before the [merge](https://github.com/NixOS/nixpkgs/commit/8fcbb0c6ed446eb855bc5c5b84d63d004def5d10) that resolved the conflicts).

**Comment by happysalada** [comment]
@mweinelt thank you for the fix!

**Comment by fabianhjr** [comment]
libei fix: https://github.com/NixOS/nixpkgs/pull/276529

**Comment by jtojnar** [comment]
PHP tests with libxml2 should be fixed by https://github.com/NixOS/nixpkgs/pull/275887

**Comment by mweinelt** [comment]
All postgresql versions are affected by libxml2 changes on aarch64-darwin. E.g. https://hydra.nixos.org/build/244436970

```
xml.c:1022:45: error: incompatible function pointer types passing 'void (void *, xmlErrorPtr)' (aka 'void (void *, struct _xmlError *)') to parameter of type 'xmlStructuredErrorFunc' (aka 'void (*)(void *, const struct _xmlError *)') [-Wincompatible-function-pointer-types]
        xmlSetStructuredErrorFunc((void *) errcxt, xml_errorHandler);
                                                   ^~~~~~~~~~~~~~~~
/nix/store/vpvhjfv5090jwdbzddw4p05i9p2j7ys2-libxml2-2.12.3-dev/include/libxml2/libxml/xmlerror.h:898:29: note: passing argument to parameter 'handler' here
                                 xmlStructuredErrorFunc handler);
                                                        ^
xml.c:1854:11: warning: cast from 'void (*)(StringInfo, const char *, ...)' (aka 'void (*)(struct StringInfoData *, const char *, ...)') to 'xmlGenericErrorFunc' (aka 'void (*)(void *, const char *, ...)') converts to incompatible function type [-Wcast-function-type-strict]
                                                           (xmlGenericErrorFunc) appendStringInfo);
                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
xml.c:3847:5: warning: cast from 'void (*)(xmlDocPtr)' (aka 'void (*)(struct _xmlDoc *)') to 'void (*)(xmlNodePtr)' (aka 'void (*)(struct _xmlNode *)') converts to incompatible function type [-Wcast-function-type-strict]
                                (void (*) (xmlNodePtr)) xmlFreeDoc : xmlFreeNode;
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
xml.c:4541:55: error: incompatible function pointer types passing 'void (void *, xmlErrorPtr)' (aka 'void (void *, struct _xmlError *)') to parameter of type 'xmlStructuredErrorFunc' (aka 'void (*)(void *, const struct _xmlError *)') [-Wincompatible-function-pointer-types]
        xmlSetStructuredErrorFunc((void *) xtCxt->xmlerrcxt, xml_errorHandler);
                                                             ^~~~~~~~~~~~~~~~
/nix/store/vpvhjfv5090jwdbzddw4p05i9p2j7ys2-libxml2-2.12.3-dev/include/libxml2/libxml/xmlerror.h:898:29: note: passing argument to parameter 'handler' here
                                 xmlStructuredErrorFunc handler);
                                                        ^
xml.c:4595:55: error: incompatible function pointer types passing 'void (void *, xmlErrorPtr)' (aka 'void (void *, struct _xmlError *)') to parameter of type 'xmlStructuredErrorFunc' (aka 'void (*)(void *, const struct _xmlError *)') [-Wincompatible-function-pointer-types]
        xmlSetStructuredErrorFunc((void *) xtCxt->xmlerrcxt, xml_errorHandler);
                                                             ^~~~~~~~~~~~~~~~
/nix/store/vpvhjfv5090jwdbzddw4p05i9p2j7ys2-libxml2-2.12.3-dev/include/libxml2/libxml/xmlerror.h:898:29: note: passing argument to parameter 'handler' here
                                 xmlStructuredErrorFunc handler);
                                                        ^
xml.c:4738:55: error: incompatible function pointer types passing 'void (void *, xmlErrorPtr)' (aka 'void (void *, struct _xmlError *)') to parameter of type 'xmlStructuredErrorFunc' (aka 'void (*)(void *, const struct _xmlError *)') [-Wincompatible-function-pointer-types]
        xmlSetStructuredErrorFunc((void *) xtCxt->xmlerrcxt, xml_errorHandler);
                                                             ^~~~~~~~~~~~~~~~
/nix/store/vpvhjfv5090jwdbzddw4p05i9p2j7ys2-libxml2-2.12.3-dev/include/libxml2/libxml/xmlerror.h:898:29: note: passing argument to parameter 'handler' here
                                 xmlStructuredErrorFunc handler);
                                                        ^
2 warnings and 4 errors generated.

File: pkgs/by-name/bi/bitbake-language-server/package.nix [hunk 1]
```diff
@@ -26,13 +26,7 @@ python3.pkgs.buildPythonApplication rec {
...
   ];
 
-  # The scikit-build-core runs CMake internally so we must let it run the configure step itself.
-  dontUseCmakeConfigure = true;
-  SKBUILD_CMAKE_ARGS = lib.strings.concatStringsSep ";" [
-    "-DFETCHCONTENT_FULLY_DISCONNECTED=ON"
-    "-DFETCHCONTENT_QUIET=OFF"
-    "-DFETCHCONTENT_SOURCE_DIR_TREE-SITTER-BITBAKE=${tree-sitter-bitbake}"
-  ];
+  SETUPTOOLS_SCM_PRETEND_VERSION = version;
```
Comment (mweinelt) [inline pkgs/by-name/bi/bitbake-language-server/package.nix:29]: This is redundant, since setuptools-scm now has a setupHook setting it.

**Comment by K900** [comment]
Probably need https://sourceforge.net/p/xmlstar/patches/22/

**Comment by vcunat** [comment]
- [ ] `nix` (2.18.1) fails on `aarch64-darwin` (only, repeatedly)
  https://hydra.nixos.org/build/245419717#tabs-buildsteps

**Comment by vcunat** [comment]
- [ ] `python3*Packages.openpyxl` fails some test (all platforms); most likely also related to libxml2 upgrade, but I failed to find more
  https://hydra.nixos.org/build/245414928/nixlog/2/tail
  _Report says this affects over 200 builds._

**Comment by leo60228** [comment]
It appears that pkgsStatic is not actually producing static binaries on aarch64-linux. I'm not sure what the root cause of this is.

**Comment by vcunat** [comment]
`pkgsStatic` should be addressed by https://github.com/NixOS/nixpkgs/pull/279097

**Comment by jtojnar** [comment]
I checked about 310 failing build logs (skipping python, kernels and haskell, one per attribute (single system representative)), these are the failures that look caused by libxml2 bump:

- [ ] [`gtkpod` x86_64-linux](https://hydra.nixos.org/build/245395608/nixlog/4)
- [ ] [`jsonnet` x86_64-darwin](https://hydra.nixos.org/build/245414573/nixlog/18)
- [ ] [`libosinfo` x86_64-darwin](https://hydra.nixos.org/build/245386492/nixlog/4)
- [ ] [`lxqt.obconf-qt` aarch64-linux](https://hydra.nixos.org/build/245626498/nixlog/3)
- [ ] [`mailcore2` x86_64-linux](https://hydra.nixos.org/build/245486701/nixlog/4)
- [ ] [`mapserver` aarch64-darwin](https://hydra.nixos.org/build/245434861/nixlog/4)
- [ ] [`mysql-workbench` x86_64-linux](https://hydra.nixos.org/build/245755555/nixlog/3)
- [ ] [`openlierox` x86_64-linux](https://hydra.nixos.org/build/245474445/nixlog/4)
- [ ] [`osinfo-db-tools` aarch64-darwin](https://hydra.nixos.org/build/245467415/nixlog/4)
- [ ] [`pdal` x86_64-darwin](https://hydra.nixos.org/build/245353431/nixlog/4)
- [ ] [`php81Extensions.dom` x86_64-linux](https://hydra.nixos.org/build/245364434/nixlog/4)
- [ ] [`php81Extensions.zend_test` x86_64-linux](https://hydra.nixos.org/build/245377983/nixlog/4)
- [ ] [`php83Packages.php-parallel-lint` x86_64-linux](https://hydra.nixos.org/build/245452993/nixlog/4)
- [ ] [`pidgin` aarch64-darwin](https://hydra.nixos.org/build/245347540/nixlog/6)
- [ ] [`precice` x86_64-linux](https://hydra.nixos.org/build/245587764/nixlog/4)
- [ ] [`pulseeffects-legacy` x86_64-linux](https://hydra.nixos.org/build/245763186/nixlog/2)
- [ ] [`rubyPackages_3_1.libxml-ruby` aarch64-linux](https://hydra.nixos.org/build/245458455/nixlog/4)
- [ ] [`rubyPackages_3_1.nokogiri` x86_64-darwin](https://hydra.nixos.org/build/245464310/nixlog/4)
- [ ] [`telepathy-gabble` aarch64-darwin](https://hydra.nixos.org/build/245478141/nixlog/4)
- [ ] [`unit` aarch64-linux](https://hydra.nixos.org/build/245391467/nixlog/45)
- [ ] [`virtualbox` x86_64-linux](https://hydra.nixos.org/build/245760105/nixlog/2)
- [ ] [`xmlcopyeditor` aarch64-darwin](https://hydra.nixos.org/build/245504931/nixlog/4)
- [ ] [`xmlsec` aarch64-darwin](https://hydra.nixos.org/build/245433949/nixlog/4)
- [ ] [`zeroadPackages.zeroad-unwrapped` aarch64-linux](https://hydra.nixos.org/build/245759050/nixlog/2)

**Comment by imincik** [comment]
@sikmir , are you able to have a look on failed build of mapserver on [aarch64-darwin](https://hydra.nixos.org/build/245434861/nixlog/4) ?

**Comment by K900** [comment]
That just looks like more libxml2 shenanigans. Needs a `const` added there.

**Comment by imincik** [comment]
> @sikmir , are you able to help with failed build of mapserver on [aarch64-darwin](https://hydra.nixos.org/build/245434861/nixlog/4) ?

Actually, https://github.com/MapServer/MapServer/pull/6971 looks like potential fix

**Comment by imincik** [comment]
@jtojnar , mapserver PR created - https://github.com/NixOS/nixpkgs/pull/279418 . Thanks for your work !

**Comment by LunNova** [comment]
notmuch is failing

```
notmuch>  FAIL   Config list from python (traditional)
notmuch>        --- T055-path-config.19.EXPECTED        2024-01-08 21:02:14.227718960 +0000
notmuch>        +++ T055-path-config.19.OUTPUT  2024-01-08 21:02:14.233718985 +0000
notmuch>        @@ -1,10 +0,0 @@
notmuch>        -database.autocommit
notmuch>        -database.backup_dir
notmuch>        -database.hook_dir
notmuch>        -database.mail_root
notmuch>        -database.path
notmuch>        -maildir.synchronize_flags
notmuch>        -new.tags
notmuch>        -user.name
notmuch>        -user.other_email
notmuch>        -user.primary_email
notmuch> Traceback (most recent call last):
notmuch>   File "<stdin>", line 1, in <module>
notmuch> ModuleNotFoundError: No module named 'notmuch2'
```

**Comment by LunNova** [comment]
https://github.com/NixOS/nixpkgs/pull/279694

**Comment by jansol** [comment]
#279885

**Comment by jansol** [comment]
And a couple more following the exact same pattern while I am at it: #279904

**Comment by yisraeldov** [comment]
Hello @vcunat This merge breaks `libresprite` build 

found with 

```
git bisect start  --first-parent origin/master b0db08979c620fc3b774fdda88db0943d723eb59
git bisect run nix-build -A libresprite `pwd`
```

**Comment by K900** [comment]
This is to some extent expected. Staging merges will always cause some amount of regressions, as it's impossible to address every single one in a reasonable timeframe.

**Comment by yisraeldov** [comment]
Thanks, so what is the right way to report this so it doesn't get forgotten?

**Comment by K900** [comment]
Fixed: https://github.com/NixOS/nixpkgs/pull/281732

Please test.

**Comment by yisraeldov** [comment]
> Fixed: #281732
> 
> Please test.

Works #:+1: 

Thank you. 

@K900 you are a scholar and a gentleman

**Pull request merged**